lib/qos/qos.c

```diff
@@ -0,0 +1,335 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <sys/slist.h>
+
+#include "qos.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(qos, CONFIG_QOS_LOG_LEVEL);
+
+#define INVALID -1
+
+/* Lookup table for QoS backoff messages. */
+uint32_t backoff_delay_lookup_sec[] = {
+	16, 32, 64
+};
+
+/* Structure used to keep track of pending messages. Used in combination with a linked list. */
+struct qos_metadata {
+	/* Mandatory variable used to construct a linked list. */
+	sys_snode_t header;
+
+	/* Message associated with each node in the linked list. */
+	struct qos_data message;
+};
+
+/* Structure containing internal variables in the library.  */
+static struct ctx {
+	/* Library event handler. Used in callbacks to the caller. */
+	qos_evt_handler_t app_evt_handler;
+
+	/* Number of times the message retransmission has been invoked by the library.
+	 * Used to index backoff delay.
+	 */
+	uint8_t retry_count;
+
+	/* Internal array of pending messages. Used in combination with linked list. */
+	struct qos_metadata list_internal[CONFIG_QOS_PENDING_MESSAGES_MAX];
+
+	/* Linked list used to keep track of pending messages. */
+	sys_slist_t pending_list;
+
+	/* Variable used to prevent multiple library initializations. */
+	bool initialized;
+
+	/* Delayed work usd to schedule a backoff timer. */
```

Comment:
I think it's more accurate to say that the work is used as the backoff timer. Based on the comment I was expecting the work item to start a timer, which it doesn't do.
```suggestion
	/* Delayed work used as the backoff timer. */
```

Reply:
True. Will apply your suggestion.